File: 2025/d10/p2.rb
require 'z3'
tot=0
File.readlines('input.txt',chomp:true).each do |l|
  parts=l.split(' ')
  btns=parts[1..-2].map {|b| b[1..-2].split(',').map(&:to_i)}
  targets=parts[-1][1..-2].split(',').map(&:to_i)
  solver=Z3::Optimize.new
  vars=btns.each_index.map {|i| Z3.Int("p#{i}")}
  vars.each {|v| solver.assert v>=0}

  targets.each_with_index do |target,c|
    affecting=btns.each_with_index.select {|btn,_| btn.include?(c)}.map(&:last)
    if affecting.any?
      s=affecting.map {|i| vars[i]}.inject(:+)
      solver.assert s==target
    end
  end

  totvar=vars.inject(:+)
  solver.minimize totvar
  if solver.satisfiable?
    model=solver.model
    tot+=vars.map {|v| model[v].to_i}.sum
  end
end
p tot

File: 2025/d10/p2.js
const fs=require('fs')
const {init}=require('z3-solver')
fs.readFile('input.txt','utf8',async (err,data)=>{
	if(err) throw err
	const ctx=await init()
	const {Optimize,Int} = new ctx.Context()
	let tot=0
	for(const l of data.trim().split('\n')){
		const p=l.split(' ')
		const btns=p.slice(1,-1).map(b=>b.slice(1,-1).split(',').map(Number))
		const targets=p.at(-1).slice(1,-1).split(',').map(Number)
		const solver=new Optimize()
		const vars=btns.map((_,i) => Int.const(`p${i}`))
		vars.forEach(v=>solver.add(v.ge(0)))
		targets.forEach((target,c)=>{
			const s=vars.filter((_,i)=>btns[i].includes(c)).reduce((acc,v)=>acc.add(v), Int.val(0))
			solver.add(s.eq(Int.val(target)))
		})
		
		solver.minimize(vars.reduce((acc,v) => acc.add(v), Int.val(0)))
		if (await solver.check()==='sat') {
			const model=solver.model()
			tot+= +model.eval(vars.reduce((acc,v)=>acc.add(v), Int.val(0))).toString()
		}
	}
	console.log(tot)
})
